### PR TITLE
finishing out rational properties

### PR DESCRIPTION
**Pull request by riz0id**
- added ring definitions for unnormalised and normalised rationals as well as all of their prerequisites. 
- added morphisms for rings
- cried because numbers make me sad 

will add/remove/fix whatever just let me know

### CODE REVIEW

File: src/Algebra/Morphism/GroupMonomorphism.agda [hunk 1]
```diff
@@ -0,0 +1,76 @@
+{-# OPTIONS --without-K --safe #-}
```
Comment (gallais) [inline src/Algebra/Morphism/GroupMonomorphism.agda]: You're missing a header here. Look at any other file in the library
and grab & modify the top banner.

File: src/Algebra/Morphism/RingMonomorphism.agda [hunk 2]
```diff
@@ -0,0 +1,87 @@
+{-# OPTIONS --without-K --safe #-}
```
Comment (gallais) [inline src/Algebra/Morphism/RingMonomorphism.agda]: Same lack of header here.

File: src/Algebra/Morphism/RingMonomorphism.agda [hunk 3]
```diff
@@ -0,0 +1,87 @@
...
+  zeroˡ zeroˡ x = injective (begin
+    ⟦ 0# * x ⟧     ≈⟨ *-homo 0# x ⟩
+    ⟦ 0# ⟧ ⊛ ⟦ x ⟧ ≈⟨ *-cong ε-homo refl ⟩
+    0#₂ ⊛ ⟦ x ⟧    ≈⟨ zeroˡ ⟦ x ⟧ ⟩
+    0#₂            ≈˘⟨ ε-homo ⟩
+    ⟦ 0# ⟧ ∎)
+
+  zeroʳ : RightZero _≈₂_ 0#₂ _⊛_ → RightZero _≈₁_ 0# _*_
+  zeroʳ zeroʳ x = injective (begin
+    ⟦ x * 0# ⟧    ≈⟨ *-homo x 0# ⟩
```
Comment (gallais) [inline src/Algebra/Morphism/RingMonomorphism.agda]: Could you align the `≈` signs here please?

File: src/Data/Rational/Properties.agda [hunk 4]
```diff
@@ -597,6 +650,15 @@ private
...
 +-0-isCommutativeMonoid = +-Monomorphism.isCommutativeMonoid ℚᵘ.+-0-isCommutativeMonoid
 
++-0-isGroup : IsGroup _+_ 0ℚ (-_)
++-0-isGroup = +-0-Monomorphism.isGroup ℚᵘ.+-0-isGroup
+
++-0-isAbelianGroup : IsAbelianGroup _+_ 0ℚ (-_)
++-0-isAbelianGroup = +-0-Monomorphism.isAbelianGroup ℚᵘ.+-0-isAbelianGroup
+
++-0-isRing : IsRing _+_ _*_ -_ 0ℚ 1ℚ
++-0-isRing = {!!}
```
Comment (JacquesCarette) [inline src/Data/Rational/Properties.agda]: There is a hole left in this PR.

File: src/Data/Rational/Unnormalised/Properties.agda [hunk 5]
```diff
@@ -556,3 +582,9 @@ p ≤? q = Dec.map′ *≤* drop-*≤* (↥ p ℤ.* ↧ q ℤ.≤? ↥ q ℤ.* 
 +-*-ring = record
   { isRing = +-*-isRing
   }
+
++-*-commutativeRing : CommutativeRing 0ℓ 0ℓ
++-*-commutativeRing = record
```
Comment (JacquesCarette) [inline src/Data/Rational/Unnormalised/Properties.agda]: I like putting 1-field record definitions on 1 line.

File: CHANGELOG.md [hunk 6]
```diff
@@ -356,6 +359,11 @@ Other major additions
 Other minor additions
 ---------------------
 
+* Added new record to `Algebra.Bundles`:
+  ```
+  record RawAbelianGroup c ℓ : Set (suc (c ⊔ ℓ)) where
```
Comment (MatthewDaggitt) [inline CHANGELOG.md]: This has been removed?

File: src/Algebra/Bundles.agda [hunk 7]
```diff
@@ -603,6 +602,20 @@ record RawRing c ℓ : Set (suc (c ⊔ ℓ)) where
     0#      : Carrier
     1#      : Carrier
 
+  +-rawGroup : RawGroup c ℓ
```
Comment (MatthewDaggitt) [inline src/Algebra/Bundles.agda]: Needs an entry in the CHANGELOG

File: src/Algebra/Morphism/Structures.agda [hunk 8]
```diff
@@ -123,9 +122,142 @@ module MonoidMorphisms (M₁ : RawMonoid a ℓ₁) (M₂ : RawMonoid b ℓ₂) w
...
+  open RawGroup G₁ renaming
+    (Carrier to A; _≈_ to _≈₁_; _∙_ to _∙_; _⁻¹ to _⁻¹₁; ε to ε₁)
+  open RawGroup G₂ renaming
+    (Carrier to B; _≈_ to _≈₂_; _∙_ to _◦_; _⁻¹ to _⁻¹₂; ε to ε₂)
+  open MorphismDefinitions A B _≈₂_
+  open FunctionDefinitions _≈₁_ _≈₂_
+  open MagmaMorphisms (RawGroup.rawMagma G₁) (RawGroup.rawMagma G₂)
+  open MonoidMorphisms (RawGroup.rawMonoid G₁) (RawGroup.rawMonoid G₂)
+
+  record IsGroupHomomorphism (⟦_⟧ : A → B) : Set (a ⊔ ℓ₁ ⊔ ℓ₂) where
```
Comment (MatthewDaggitt) [inline src/Algebra/Morphism/Structures.agda]: This and the other records below need an entry in the CHANGELOG

**Comment by riz0id** [comment]
sorry for all the issues, hopefully that's the last of them. thanks for taking the time to look over this one.

**Comment by MatthewDaggitt** [comment]
> sorry for all the issues, hopefully that's the last of them. thanks for taking the time to look over this one.

Haha, no we're the ones that should be thanking you for the great contribution :+1:

**Pull request merged**